Added helm OCI registry support

itsfarhan opened the pull request:
Previously, we had issue of blast radius while setting up gh-pages.. basically docs updates could break chart serving.
Therefore in #1979 #1985 chart preservation was done in `.github/workflows/gh-pages.yaml` temporarily

But as mentioned by @ytsarev in #1973 as a good possible solution. I have done following this to implement this solution.

- Add OCI registry publishing (`oci://ghcr.io/k8gb-io/charts/k8gb`)
- chart preservation logic from gh-pages workflow also exist.

**Testing:**
- [x] Verify OCI registry publishing works on release v0.15.0
- [x] Test installation methods work

**Steps done:**
- `helm dependency update chart/k8gb` and `helm package chart/k8gb` -> successfully created k8gb-v0.15.0.tgz
- `echo $GITHUB_TOKEN | helm registry login ghcr.io -u itsfarhan --password-stdin` and `helm push k8gb-*.tgz oci://ghcr.io/itsfarhan/charts` ->  Chart successfully pushed to ghcr.io/itsfarhan/charts/k8gb:v0.15.0
- `helm install k8gb oci://ghcr.io/itsfarhan/charts/k8gb --version v0.15.0 -n k8gb --create-namespace` -> Chart pulled and installed successfully.
- Tested on different branch and docs only deployed without chart files. (no preservation needed)

<img width="2043" height="587" alt="Screenshot from 2025-08-09 00-36-27" src="https://github.com/user-attachments/assets/42d5f90e-e0a9-43a5-8164-4ba104c3d052" />
<img width="2059" height="375" alt="Screenshot from 2025-08-09 00-02-15" src="https://github.com/user-attachments/assets/5229abb4-4014-4605-98d2-91ca60414cea" />

<details>
  <summary>HOW TO RUN CI</summary>
---

  By default, all the checks will be run automatically. Furthermore, when changing website-related stuff, the preview will be generated by the netlify bot.

  ### Heavy tests
  Add the [`heavy-tests`](/k8gb-io/k8gb/issues?q=is%3A*+label%3Aheavy-tests) label on this PR if you want full-blown tests that include more than 2-cluster scenarios.

  ### Debug tests
  If the test suite is failing for you, you may want to try triggering `Re-run all jobs` (top right) with [debug logging](https://docs.github.com/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging) enabled. It will also make the [print debug](/k8gb-io/k8gb/blob/master/.github/actions/print-terratest-debug/action.yaml) action more verbose.

</details>

abaguas commented on .github/workflows/helm_publish.yaml:
```diff
@@ -19,17 +19,22 @@ jobs:
+      - name: Push to OCI Registry
+        run: |
+          echo ${{ secrets.GITHUB_TOKEN }} | helm registry login ghcr.io -u ${{ github.actor }} --password-stdin
+          helm push k8gb-*.tgz oci://ghcr.io/k8gb-io/charts
```
I am not sure if I follow. Where is the .tgz file created?

I quickly read through the docs of the job you commented out, and helm package seems to be run there: https://github.com/dave-mcconnell/helm-gh-pages-microservices

abaguas commented on .github/workflows/gh-pages.yaml:
```diff
@@ -48,9 +48,7 @@ jobs:
           git fetch origin gh-pages:gh-pages
           if git checkout gh-pages; then
             echo "Backing up existing files..."
-            test -d charts && cp -r charts /tmp/preserved/
```
Why did these files need to be preserved?
I quickly read https://github.com/k8gb-io/k8gb/pull/1985 but I couldn't understand.

k0da replied:
IMO we should keep those since we push to oci only starting with v0.15/v0.16. For backward compatibility we should keep standard helm repo

itsfarhan replied:
sure @ytsarev

itsfarhan commented:
@ytsarev 

we can make Helm registry discovered by artifacthub. We need to add OCI url in artifacthub-repo.yml
And also add OCI registry in artifacthub - https://artifacthub.io/control-panel/ 

artifacthub-repo.yml metadata:

```bash
repositoryID: caa5085b-79f6-42b4-848e-a3c6c202066b
name: k8gb
url: oci://ghcr.io/k8gb-io/charts
```

References:
https://github.com/artifacthub/hub/blob/master/docs/metadata/artifacthub-repo.yml
https://github.com/artifacthub/hub/blob/master/docs/helm_charts_repositories.md

ytsarev reviewed (commented):
Currently we have https://github.com/k8gb-io/k8gb/blob/gh-pages/artifacthub-repo.yml, we will have to switch/amend it to point to OCI

ytsarev commented on chart/k8gb/README.md:
```diff
@@ -138,7 +138,7 @@ For Kubernetes `< 1.19` use this chart and k8gb in version `0.8.8` or lower.
 | k8gb.securityContext.runAsUser | int | `1000` |  |
 | k8gb.serviceMonitor | object | `{"enabled":false}` | enable ServiceMonitor |
-| k8gb.tolerations | array | `[]` | Tolerations to apply to the k8gb operator deployment |
+| k8gb.tolerations | list | `[]` | Tolerations to apply to the k8gb operator deployment for example:   tolerations:   - key: foo.bar.com/role     operator: Equal     value: master     effect: NoSchedule |
```
does not look related to this PR?

itsfarhan replied:
Oh, no this is not related to this PR. It must be changes from the upstream repository I pulled. I will remove anything not related to this PR and commit again.

ytsarev commented on .github/workflows/helm_publish.yaml:
```diff
@@ -19,6 +19,7 @@ jobs:
       - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
         with:
           fetch-depth: 0
+      # TODO: Adding dave-mcconnell/helm-gh-pages-microservices
```
Some leftover comment?

ytsarev reviewed (commented):
@itsfarhan let's remove unrelated to the PR semantic changes, merge the registry push and publish to OCI in parallel for a while

itsfarhan commented:
> @itsfarhan let's remove unrelated to the PR semantic changes, merge the registry push and publish to OCI in parallel for a while

@ytsarev I have removed unrelated changes and added helm charts preservation and oci related configs in helm publish yaml file.

ytsarev commented on .github/workflows/gh-pages.yaml:
```diff
@@ -58,6 +58,9 @@ jobs:
       - name: Deploy to gh-pages
         run: |
+          # Copy built site to temp location
+          cp -r site /tmp/mkdocs-site
```
@itsfarhan do we use it anywhere? can't find ` /tmp/mkdocs-site` in the rest of the script

itsfarhan replied:
You're right. I missed updating deployment path to use `/tmp/mkdocs-site/*` when I restored the preservation logic.

I have referred to your previous commits on pr - #1987 and #1985 to restore things

ytsarev replied:
Can you rebase with `master` branch please ? I already see that changes there https://github.com/k8gb-io/k8gb/blob/master/.github/workflows/gh-pages.yaml#L61-L85 Something is wrong with the state of your feature branch

itsfarhan replied:
Sure @ytsarev 
Done! Rebased with master branch. Everything is back to its place and OCI support added.

ytsarev reviewed (approved):
Thanks a lot for the contribution
